### ISSUE TITLE
--insecure-registry should be on "docker pull"

### ISSUE THREAD

**Issue by abourget** [docker/docker#8887]
Hi folks, thanks for all your great work.

I was previously running a "library/registry" on `localhost:5000`. With Docker 1.3+, I was required to run *docker* with `--insecure-registry localhost:5000`.  Doing so did nothing, until I discovered I needed to run **docker**, as in **daemon**, with those parameters.

It would be very useful to have that handled directly by `docker pull`, and not have to restart the whole thing and tweak system-level settings when you discover you need to use a local unsecure registry.  **EDIT:** As mentioned in the comments, it would also be very useful to allow *any* registry to be unsecure, not just named ones, as Docker sometimes provides random ports, and some environments have many registries popping in and out of existence.

It is currently read here: https://github.com/docker/docker/blob/master/docker/daemon.go#L43 (while running the daemon), and it's checked while `pull`ing in https://github.com/docker/docker/blob/master/graph/pull.go#L116 .. maybe we could add yet another switch to `pull` like `--insecure` and tweak that would forcefully make it `secure == false` ?

I don't have a docker development setup ready, but if you think it's a good idea.. I could try to implement it.

```
Linux cerise 3.13.0-32-generic #57-Ubuntu SMP Tue Jul 15 03:51:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
Docker version 1.3.1, build 4e9bbfa
Containers: 5
Images: 607
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Dirs: 618
Execution Driver: native-0.2
Kernel Version: 3.13.0-32-generic
Operating System: Ubuntu 14.04.1 LTS
Debug mode (server): false
Debug mode (client): true
Fds: 10
Goroutines: 11
EventsListeners: 0
Init Path: /usr/bin/docker
Username: abourget
Registry: [https://index.docker.io/v1/]
WARNING: No swap limit support
```

**Comment by djannot** [comment]
+1

**Comment by maelp** [comment]
+1

**Comment by ryan-stateless** [comment]
+1

**Comment by jonathanvaughn** [comment]
+1

**Comment by jaytaylor** [comment]
+1, this is currently difficult to get configured and working.

@mhamrah makes great points.  Don't force things, let user's decide and configure for their own needs.

**Comment by mhamrah** [comment]
Registries that use self-signed certificates are also a problem, especially
in a developer context using boot2docker which switched to a read-only file
system. This requires an additional and different configuration step to
bootstrap the running vm.

I believe everyone that's posted to this thread sees the value and benefit
of docker, are using it daily, are trying to promote it in their
organizations, but are experiencing a painful and unnecessary issue that
hinders adoption.

As much as we all know docker it is still unknown to many people in tech-
especially within the enterprise. Documentation helps, but we are still
jumping through hoops, and it's a big blocker with a negative overall
effect.

**Comment by NoumanSaleem** [comment]
+1

**Comment by ChiChou** [comment]
:+1:

**Comment by jsdir** [comment]
:+1:

**Comment by tangr1** [comment]
+1

**Comment by toddbu** [comment]
Super disappointing that we still can't pull from insecure registries without passing a flag to the docker daemon.  This is just another hassle for every new machine we deploy.

**Comment by aidanhs** [comment]
+1

Some thoughts:
1. could we have hostname wildcards? e.g. `--insecure-registry=*.internal`
2. could self-signed certs be treated differently to http?
3. related to 2, could docker do something similar to SSH and prompt the user to accept a self-signed cert whenever it sees a new one/complain loudly if there's a mismatching one?

**Comment by aidanhs** [comment]
And while I'm making suggestions, why not treat localhost as always being secure? (as Chrome does)

**Comment by agquick** [comment]
+1000 on this.. and +1 on the configuration reload feature, that's what makes this twice as bad. I stayed with v1.2 because I figured surely a maintainer would realize how annoying the --insecure-registry flag on the daemon is for deployment and fix it, but somehow the minor releases keep ignoring it.

For example, if I had to change my private registry IP for whatever reason, I would have to restart every single docker daemon on every VM -- just because my registry changed!? Those 2 things should never be coupled that tightly. And telling people to just add 0.0.0.0/8 defeats the whole purpose of the security implementation in the first place.

Adding this flag to the push/pull commands seems so obvious as a fallback for the daemon flag, please someone explain to me why they are still fighting it but adding "nice to have" features in the meantime.

**Comment by seanmcl** [comment]
@agquick's comment is spot on, especially wrt "nice to have" features.

**Comment by tiborvass** [comment]
Realizing this is a continuing pain for a substantial number of users, we are reconsidering adding `--insecure` to `pull`. @ewindisch and I will be working on a PR that we'll link to this issue.

Apologies for the long wait, and thanks for voicing your opinions and pointing out your pain points.

**Comment by thaJeztah** [comment]
@tiborvass I can imagine there's an equally large number of users that *don't* want to allow pulls from insecure registries. I realise Docker currently doesn't have fine-grained control over permissions/config, but if there's a chance to be able to "lock" this down, I think that would be a nice touch.

**Comment by octalthorpe** [comment]
Oh my make it so! Was about to implement it myself.

Sent from my BlackBerry 10 smartphone on the Bell network.
From: Sebastiaan van Stijn
Sent: Monday, February 23, 2015 4:53 PM
To: docker/docker
Reply To: docker/docker
Cc: Kristopher Cieplak
Subject: Re: [docker] --insecure-registry should be on "docker pull" (#8887)



@tiborvass<https://github.com/tiborvass> I can imagine there's an equally large number of users that don't want to allow pulls from insecure registries. I realise Docker currently doesn't have fine-grained control over permissions/config, but if there's a chance to be able to "lock" this down, I think that would be a nice touch.

**Comment by aidanhs** [comment]
@thaJeztah I'm trying to figure out what use case you're thinking of that means we can't have a `--insecure-registry` flag to `docker pull`.
 - if a user doesn't want to be accidentally MITMed on a secure registry, they can just avoid passing `--insecure-registry`
 - if a user wants to enforce on users that all images come from secure registries (i.e. an 'enterprisey' scenario), they cannot actually enforce this at all at the moment anyway!

So could you elaborate on what `docker pull --insecure-registry` inhibits?

----

To elaborate on my second point, I can't see how you'd lock this down without rethinking a significant amount of how docker works! Putting aside being able to `docker load` a tarball you could obtain by writing your own registry puller, and using `docker run -v` to escalate privs and add something to the daemon arguments, there's a very simple way to bypass any 'enforcement':
```
$ docker pull registry:5000/aidanhs/blah                                    
FATA[0004] Error: v1 ping attempt failed with error: Get https://registry:5000/v1/_ping: EOF. If this private registry supports only HTTP or HTTPS [...]
$ socat tcp4-listen:5000,reuseaddr,fork tcp4:registry:5000 &
[1] 22924
$ docker pull localhost:5000/aidanhs/blah
Pulling repository localhost:5000/aidanhs/blah
[...]
Status: Image is up to date for localhost:5000/aidanhs/blah:latest
$ docker tag localhost:5000/aidanhs/blah registry:5000/aidanhs/blah
```

**Comment by thaJeztah** [comment]
> if a user wants to enforce on users that all images come from secure registries (i.e. an 'enterprisey' scenario), they cannot actually enforce this at all at the moment anyway!

This is the scenario, yes.

> To elaborate on my second point, I can't see how you'd lock this down without rethinking a significant amount of how docker works

I fully understand that (at this moment) there's no way to lock it down. Users having access to `docker.sock` have effectively root access, so can change anything they like. Also, they would still be able to change the daemon flag and restart the daemon.

However, it *does* give a clear signal to the user if `docker pull --insecure-registry ..` gives an error ("Insecure registries are disabled"), which *would* notify users that it isn't desired, for example, when trying some example they found somewhere.

So, would it cover all cases? No. Would it *hurt*, I don't think so either.

**Comment by aidanhs** [comment]
I personally feel it'd do more harm than good, simply because it misleads people to think "oh look, docker enforces this" when actually it's a very superficial protection. I could go on, but in the end I think it comes down to a matter of taste.

**Comment by justinclayton** [comment]
If you're looking to how it might hurt, just scroll up. There are myriad UX issues with this approach that create barriers to adoption, and they're all detailed above. 




On Mon, Mar 2, 2015 at 2:03 AM -0800, "Sebastiaan van Stijn" <notifications@github.com> wrote:













if a user wants to enforce on users that all images come from secure registries (i.e. an 'enterprisey' scenario), they cannot actually enforce this at all at the moment anyway!




This is the scenario, yes.




To elaborate on my second point, I can't see how you'd lock this down without rethinking a significant amount of how docker works




I fully understand that (at this moment) there's no way to lock it down. Users having access to docker.sock have effectively root access, so can change anything they like. Also, they would still be able to change the daemon flag and restart the daemon.



However, it does give a clear signal to the user if docker pull --insecure-registry .. gives an error ("Insecure registries are disabled"), which would notify users that it isn't desired, for example, when trying some example they found somewhere.



So, would it cover all cases? No. Would it hurt, I don't think so either.

**Comment by Frusty** [comment]
+1

**Comment by sunyc** [comment]
I view the issues mostly with docker unable to restart without killing sub
containers. that make configuring/reconfiguring it much harder.

**Comment by jdoliner** [comment]
I'm pretty disappointed that there hasn't been some action taken on this issue yet. It's obviously causing a number of people pain.

It's obviously annoying to a huge number of people, it's actively blocking my work right now. Having to restart the Docker daemon to allow pulling from an insecure registry ranges from annoying to downright impossible depending on the situation you're in.

The main argument for not doing this seems to be that the system administrator should be able to lock the system down and make sure that only images from secure repos can be pulled. That use case is definitely real but I think it makes a bad default. It seems a lot more reasonable to have a flag that can be passed to the daemon on start like `--no-insecure` which disables `--insecure-registry` usage in `pull`. That way an admin can lock things down if s/he really wants to.

**Comment by ewindisch** [comment]
For those highly desiring this behavior, I do offer the following workaround. I understand it will not be viable for all users.

See my project 'docker-pull': https://github.com/ewindisch/docker-pull

You would use it as such:
```
docker run ewindisch/docker-pull --insecure-registry 10.0.0.0/16 10.0.1.2/someimage | docker load
```

It's also possible to allow all registries as insecure:
```
docker run ewindisch/docker-pull --insecure 10.0.1.2/someimage | docker load
```

I'll remind that it is *still* highly ill-advised to actually do this.

**Comment by jdoliner** [comment]
@ewindisch nifty hack.

Another pretty good solution is to use a tcp tunnel:

```shell
$ docker pull host:5000/image #fails
$ ssh -N -L 5000:host:5000 user@host
$ docker pull localhost:5000/image #works
```

**Comment by justinclayton** [comment]
These are both fantastic workarounds!





On Thu, Apr 16, 2015 at 1:43 AM -0700, "Joe Doliner" <notifications@github.com> wrote:












@ewindisch nifty hack.



Another pretty good solution is to use a tcp tunnel:

$ docker pull host:5000/image #fails
$ ssh -N -L 5000:host:5000 user@host
$ docker pull localhost:5000/image #works

**Comment by jaytaylor** [comment]
I too would like to see the default reversed.

**Comment by tobegit3hub** [comment]
+1

**Comment by testaccountspivey** [comment]
+1

**Comment by echox** [comment]
so this is reopened and the current status after 4 months is nothing and as a workaround use a bunch of hacks? :-/ Is there some discussion about this going on elsewhere or is it just dead?

And yes, +1

**Comment by devsli** [comment]
+1

**Comment by AngryHacker** [comment]
+1

**Comment by CleanCut** [comment]
I would like to point out that I counted the instances of "+1" on this page, and the count comes to 31 so far.  That's not counting the other supportive comments that don't actually include that exact string.

**Comment by cpuguy83** [comment]
The issue here is that enabling `--insecure` on pull takes the control away from the sysadmin, where it belongs.
Security is hard, and making changes to loosen security is not a small decision.
Also, people that are perfectly happy with the current setup are not going to come on here and `-1` this either.
On the other hand...
It is trivial to re-configure the daemon to enable insecure pulls from a registry.
It's also trivial to setup some self-signed certs and not even have to re-configure docker.

**Comment by CleanCut** [comment]
"Sysadmin" is one use-case for docker, but I "developer" and "non-sysadmin" are equally valid use cases.  For a developer, providing an `--insecure` option reduces friction in the workflow.

Perhaps we could have the best of both worlds by providing an option that sysadmins could specify to *deny* the usage of an `--insecure` option.  That way, sysadmins still have full control, but non-sysadmin cases don't have to deal with the workflow pain.

What is trivial for a sysadmin can be surprisingly cumbersome for non-sysadmins.  I've had to help nearly two dozen developers make (and re-make) this configuration change across 5 different OS's used in our development group.  I actually maintain a script to automate the config change for our environments.

Our sysadmins will not currently set up self-signed certs for our registry, whether it's trivial to do or not.

Anyway, even if this doesn't change, eventually people will adapt to them.  I suppose the pain I'm dealing with will all go away the next time our sysadmins do maintenance on the registry, because at that point we should be able to get actual SSL certificates installed.  Perhaps that's the whole point.  Make it easier to take the secure path than the insecure path going forward.

**Comment by andrask** [comment]
:+1:  @CleanCut, and everyone else said it all. I only work with the developer case and reconfiguring the the docker daemon is just waste of time for us.

**Comment by ewindisch** [comment]
If you have access to the docker socket today, you *are* a sysadmin. You're
root already, you have "docker load" and already have workarounds to do
insecure registry pulls. Adding the image option to the client would not be
any less secure than the status quo.

There is something, however, to be said about intentionally increasing the
friction for developers trying to do the wrong thing in other to entice
them into doing the right one.

**Comment by thaJeztah** [comment]
@ewindisch @tiborvass reading back, I see https://github.com/docker/docker/issues/8887#issuecomment-75638745;

> Realizing this is a continuing pain for a substantial number of users, we are reconsidering adding --insecure to pull. @ewindisch and I will be working on a PR that we'll link to this issue.
> Apologies for the long wait, and thanks for voicing your opinions and pointing out your pain points.

Is that still the current position? (I don't think a PR was created)

**Comment by eddwardo** [comment]
+1

**Comment by sleaze** [comment]
This continues to bother and annoy me on a daily basis.

:( sad conehead.

**Comment by ppiersonbt** [comment]
--inscure makes complete sense from a developer POV.  Its on the person implementing docker in their environment to make it secure.

**Comment by tony-kerz** [comment]
+1

**Comment by tim-kretschmer-c2fo** [comment]
:+1:

**Comment by gollawil** [comment]
+1

**Comment by tjadams** [comment]
+1

**Comment by thrap** [comment]
+1

**Comment by halvorgb** [comment]
+1

**Comment by devsli** [comment]
+1

**Comment by mhamrah** [comment]
If this issue is still open by Halloween I think all the +1's should have a one year drown-in-our-docker-sorrows anniversary party.

**Comment by jaytaylor** [comment]
+1 for the party of sorrow!

**Comment by sroebuck** [comment]
+1

**Comment by cornfeedhobo** [comment]
+1 for the party of sorrow

**Comment by nicornk** [comment]
+1

**Comment by sebadiaz** [comment]
+1

**Comment by rsmoorthy** [comment]
There are a couple of workarounds, including using SSH tunnels -- but that requires ssh accounts on the registry host. The following workaround, does not need that.

Run the registry port forwarder, as follows:

```
docker run --name registry_forwarder -d -p 5000:5000 -e REGISTRY_HOST="<registry_host_ip>" -e REGISTRY_PORT="5000" rsmoorthy/registry-forwarder
```

And then pull or push images from your local registry:
```
docker pull localhost:5000/your-image
docker push localhost:5000/my-image
```

**Comment by justinclayton** [comment]
@rsmoorthy That is fantastic. Thank you for succinctly demonstrating the futility of this current restriction. 

Docker, please re-include this particular battery.

**Comment by JasonHartmann** [comment]
+1

**Comment by Supernomad** [comment]
I have to say that forcefully encouraging security on the user has seriously swayed my thoughts on using docker. I understand completely that we can add the --insecure-registry flag to the daemon on start, and I won't get into all the reasons that doesn't work all the time, or can be impossible.

I have a question for the Docker developers, do you guys really think that you know what is best for the end user? I for one do not require SSL on my registries at all, because the network they run on is already fully encrypted. So why on earth would I encrypt encrypted traffic, is this really doing anything besides adding overhead and moving pieces to an already complex and massive system?

Also are there really use cases where people are using a private registry over the internet? What does one do about authentication in that sense? Couldn't one just pull an image down and then tear it apart? Instead of trying to intercept it en route to another computer?

TL;DR +1

**Comment by Koriit** [comment]
+1

**Comment by mingfang** [comment]
@Supernomad well said.

**Comment by belooussov** [comment]
See it from docker's side: it is one issue that is better kept never officially implemented, but with a lot of possible workarounds to ease the pain. Some users screaming loudly is still less painful than the competition's marketing labeling docker as "insecure on purpose", and damaging its reputation forever.
Having said that, +1.

**Comment by justinclayton** [comment]
@tiborvass @ewindisch This issue is over a year old. It's been over 8 months since you said it was to be re-evaluated. Have you evaluated it? If so, what has been decided? Don't leave a brother hanging! :-)

Since this issue was initially opened and closed and reopened, the community has come up with numerous ways to fix this themselves, mainly to demonstrate the futility of this default setting:

* `ssh -N -L 5000:host:5000 user@host && docker pull localhost:5000/lolsecurity`
* `socat tcp4-listen:5000,reuseaddr,fork tcp4:registry:5000 && docker pull localhost:5000/lolsecurity`
* `docker-machine create -d virtualbox --engine-insecure-registry 0.0.0.0/0 lolsecurity`
* `docker run --name registry_forwarder -d -p 5000:5000 -e REGISTRY_HOST="<registry_host_ip>" -e REGISTRY_PORT="5000" rsmoorthy/registry-forwarder && docker pull localhost:5000/lolsecurity`

Not to mention CoreOS now ships with `--insecure-registry 0.0.0.0/0` as the default. These examples clearly show that the idea that there are separation-of-concerns lines between the "sysadmin" and the "developer" is now largely arbitrary and spurious in 2015. In fact, the very idea of containers (of which Docker is their chief evangelist) has greatly accelerated this trend away from the traditional ops that still bothers to call anyone "sysadmin" in the first place.

Anyway, I'd love to see this get closed for good, one way or another.

**Comment by jhinds** [comment]
+1

**Comment by tdensmore** [comment]
+1

**Comment by jbye** [comment]
+1

**Comment by jrosadohp** [comment]
+1

**Comment by politician** [comment]
This thread looks like a classic "swept under the rug" issue.  I'm glad to learn that 15 months after this issue was reported Docker PR's need to meddle with how I operate an internal registry on closed network still rules the engineering discussion.

Don't back down.  That's just what they _want_ you to do.  Can you even imagine how much offering a functional registry would cannibalize your ability to monetize Docker Hub?

**Comment by politician** [comment]
Why should I trust Docker Registry with my SSL private key?

For the other CoreOS users left out to dry by this meddling,
https://coreos.com/os/docs/latest/registry-authentication.html#using-a-registry-without-ssl-configured

**Comment by Supernomad** [comment]
@politician couldn't have said it better myself. It most definitely, seems like docker is just ignoring the fact that a large portion of their users are at the very least dissatisfied.

The fact is I am in the process of migrating completely away from docker and I couldn't be happier with the decision. I am currently using git and lxc and its not only hand down faster than docker it actually allows me to do what is best for my company and not what another company thinks, albeit completely and utterly incorrectly, is best for me. 

I highly suggest taking a look at the alternatives which are honestly better than docker, such as rocket, raw lxc, qemu (not the same but still better) just to name a few.

I will be strongly suggesting this course of action to everyone I know that is considering containers from now on. At the very least until the people at docker realize they have no idea, and I stress *absolutely* no idea, what is best for the end user in terms of security.

**Comment by politician** [comment]
I gave up and bought a cheap dedicated SSL cert. The Docker CLI dependency on the CA system is just too stong - self-signed certificates are not just operational frustrating, they simply don't work.

- docker-machine deletes your ca.crt override between down/up
- none of your base images include the cert making tools like drone for CI impossible 
- docker CLI uses a non-standard folder for certs, so that's just something else to remember. 
- even if you do get it to work 18 hours later, you'll always have a nagging feeling that something else is broken

Bottom line: Docker inflicts a persistent on-going penalty to your ops team when attempting to use a self-signed cert.

This is all the more frustrating since `curl -k` has existed for eternity.

**Comment by stephenenelson** [comment]
+1

**Comment by buehler** [comment]
It's pretty sad, that they don't give a damn about this. If some dev just want's to fiddle around with docker and host his own environment, normally you don't want to mess around with SSL certs and stuff. Additionally, if you have your own server at home that is nowhere near to be public, you simply don't need SSL.

**Comment by thaJeztah** [comment]
@buehler you can add the `--insecure-registry` option to your daemon's settings, or to your [daemon.json](https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file) configuration file; then you only need to configure it once, and pull without having to specify the flag

**Comment by politician** [comment]
Just an update from us: We've since stripped Registry out of our infrastructure and reverted back to using [our custom fork of dogestry with Azure Blob storage support](https://github.com/itzamna314/dogestry).  We discovered that Docker Registry was duplicating layers and our machines were running out of disk space which resulted in outages.  Registry turned out to be a time-wasting dead-end for us.

**Comment by andrask** [comment]
@buehler just to concretize @thaJeztah 's proposal, add this line to the configuration:

    --insecure-registry 0.0.0.0/0

You will be able to access whatever registry you start. Works well for testing stuff.

@politician the duplication happens if the images are tagged with different repositories. The lack of deletion of blobs is a significantly bigger issue.

**Comment by justinclayton** [comment]
@thaJeztah if it's that easy to do, but 80% (admittedly arbitrary number) of users need to do it, then  please just make it the default.

**Comment by thaJeztah** [comment]
@justinclayton no; setting that option basically says "ignore any issue with secure communication to the registry", so allow man-in-the-middle attacks "out of the box", or even the daemon falling back to non-TLS "http://". Docker *does* already set it as a default for [registries in the `127.x.x.x` range](https://github.com/docker/docker/blob/4a7bd7eaef0038b006cf23adface84081e5aab3b/registry/config.go#L85-L90).

If you have a local registry and don't want to generate a certificate for it, adding a  `--insecure-registry` to your daemon options takes less than a minute to do. It should be a deliberate action though, not something that is set by default.

**Comment by justinclayton** [comment]
@thaJeztah I understand your argument, but that can't just be the end of it. There is a significant gap in user experience for new developers as a result of this being on the daemon side. The *majority* scenario where this is a pain is a new developer who follows the instructions on docker.com to download and run the Docker Toolbox installer on their Mac. Upon completion, they immediately try to run `docker pull <internally-signed-registry>/foo` and are greeted with an error. This is the true issue. Perhaps that means this issue should be renamed?

There are lots of other ways this could also be solved for; I'm sure the community and the company could agree on one of them:

* The current name of this issue is '--insecure-registry should be on "docker pull"'. As this issue is still open, I have to assume this option is still being considered.
* If a single-command solution to this was provided (and documented) that was simple for newbie users, that would solve for this.
* In our case, the registry *is* secured. It uses a cert signed by our internal CA like everything else in our build environment. This is sufficient security. If the daemon somehow was able to honor MacOS's certificate store, that would make this headache go away as well.
* If they were prompted to add the cert or make this configuration to the docker-machine during the Toolbox install process, that would probably be ok too.

Please let the 70+ participants of this issue know what direction you plan on taking with this, or else please just close the issue so we're not left hanging. Thanks.

**Comment by apakhomov** [comment]
@thaJeztah 
There is no way to add single --insecure-registry to daemon opts without restarting all running containers. That is one of the main problems. We can't reload configuration without restart (issue lasts from 2013), we can't pull images from other insecure registries without adding option to daemon and also restart. And nowadays we still don't have a clear roadmap of fixing that issue.

**Comment by thaJeztah** [comment]
@apakhomov guess it could be added to the list of configuration changes that can be updated without restarting https://docs.docker.com/engine/reference/commandline/daemon/#configuration-reloading. Can you open a separate issue for that?

**Comment by Tails** [comment]
+1. 

Some PaaS providers don't give access to the daemon and run a private network for the user (Jelastic for example).

**Comment by kkorada** [comment]
is it possible to add multiple insecure registries to docker ?
something like **--insecure-registry x.x.x.x:xxxx --insecure-registry z.z.z.z:zzzz**

**Comment by mingfang** [comment]
@kkorada ```--insecure-registry=0.0.0.0/0``` will make Docker behave like before.

**Comment by thaJeztah** [comment]
@kkorada yes, you can specify multiple registries (the flag is `--insecure-registry=[]` - square brackets indicate it can be specified multiple times).

Also for docker 1.12, we'll make this option configurable in the `daemon.json` configuration file, without restarting the daemon. See the open pull request here: https://github.com/docker/docker/pull/22337

**Comment by kkorada** [comment]
thank you @mingfang and @thaJeztah

**Comment by derekmahar** [comment]
As @mhamrah and @justinclayton [suggested](https://github.com/docker/docker/issues/8887#issuecomment-66448575), I would also recommend a solution using ssh in addition to https, similar to how Git allows access to a respository using both ssh and HTTPS.  This might make use of the ssh workaround that @justinclayton [listed](https://github.com/docker/docker/issues/8887#issuecomment-167914876).

**Comment by mattmb** [comment]
It's mad that this isn't a client setting. Obviously it shouldn't be on by default because that would encourage insecure practice. Putting the setting on the daemon however makes it very impractical for debugging purposes or in local development environments.

My current use case: running a development environment with docker compose. The environment needs a local docker registry. It is intended to run entirely on a local VM.

The docker way: explaining how to configure docker-machine to enable insecure registry on each developers machine, possibly requiring them to recreate their docker-machine if they've already got one or edit the config by hand.

The hacky solution: socat running in each container that needs to use the registry, redirecting to localhost.

Not exactly making things easy...

**Comment by luchesar** [comment]
+1

Really a pity for --insecure not to be a client configuration!
That  creates a lot of pain for us as well very similar to many of the explanations above.
Please set --insecure-registry=0.0.0.0/0 by default pr provide a way to pass it to the docker command as well as docker-compose configurations.

**Comment by zhao-ji** [comment]
+1

**Comment by GongT** [comment]
+1

**Comment by jaytaylor** [comment]
Is it time for the annual +1 Pity Party again?

**Comment by ajhodges** [comment]
+1, if having a client-side `--insecure-registry` flag is not an option, there should be a way to specify a trusted certificate to be used for `docker pull` and `docker push`. Not everyone has access to trust certificates on a OS-level, or access to play around with the Docker daemon.

Hosted CI servers (that are being used to build docker images and push to a private registry) are a great example of when you might not have that level of access.

**Comment by ericpulvino** [comment]
+1

**Comment by tiborvass** [comment]
@ajhodges https://docs.docker.com/registry/insecure/#using-self-signed-certificates

**Comment by ajhodges** [comment]
@tiborvass
Only works if you have sudo access...

**Comment by zadam** [comment]
+1 good way to discourage devs from using docker

**Comment by PhilippHeuer** [comment]
+1, i would like to see --insecure-registry for `docker login` too.

I don't even see this as a security issue, since you are consciously doing it. And if someone unauthorized got root access this won't provide any security either. What merit does this restriction actually have?

If you had bad intentions you could just upload your *malicious docker image* to the DockerHub, which is trusted.

**Comment by navels** [comment]
+1 and frankly this boggles the mind

**Comment by Ichimonji10** [comment]
+1 to the pain train.

**Comment by harindaka** [comment]
Plus F*cking One!

**Comment by jaytaylor** [comment]
Is there a PR open for this?

**Comment by Ichimonji10** [comment]
Not AFAIK. The auto-links above are generated because some of my own code references this issue. I'll pull that reference for a while to quiet down the spam here.

**Comment by waldman** [comment]
+1 as well...

**Comment by mmchen** [comment]
+1

**Comment by eromoe** [comment]
+1 ,  it is very inconvenient to add setting to `deamon.json` and restart docker.

Different machine has different os ,  some install from `yum` ,  some directly using binary .  So I have to detect that and restart dockerd correctly ....  that 's a disaster

I just stress docker pull  need  `--insecure-registry`  flag !

**Comment by r-chris** [comment]
It's only been three years - let's not loose hope now

**Comment by justinclayton** [comment]
bump 🤣

**Comment by schwaizi** [comment]
+1

**Comment by sherif84** [comment]
+1

**Comment by mmpfeffer** [comment]
The argument that this encourages users to set their registries as secure is not only presumptuous, it also missing a key point.  It puts operations in the position where many people must have access to the root account in order to do their operations work, where docker registries are moving around a lot. 
That coupling, from an operational security standpoint, creates greater security risk.

Secondly, in a private network (such as in a multi-tiered cloud-hosted application), the securing of registries is not necessary, and further complicates technology implementations which sit on top, requiring layers of security management (to handle automated docker authentication/refresh) wherever a secure docker registry is used.

At the very least, the docker daemon should be configurable to *ALLOW* insecure registry parameter to be passed in. That moves the security design to the proper place - in the hands of the administrator, and outside of docker itself.

**Comment by mjambon** [comment]
FWIW I'd be happy with a command to "add some registry to the list of insecure registries", since patching json configs from shell scripts is a major pain point.

**Comment by dmwoods38** [comment]
+1

**Comment by TintypeMolly** [comment]
:+1: +1

**Comment by sakethkoduru** [comment]
+1

**Comment by yashpatel5400** [comment]
+1

**Comment by zulrang** [comment]
+1

**Comment by csegedicsaba** [comment]
+1

**Comment by ewindisch** [comment]
Many of the requested implementations of this would make it impossible for companies utilizing Docker to adhere to SOC compliance requirements, etc.

There should be a solution to this, but I don't think there's an easy solution that will not be a more drastic architectural change to how images are stored and executed.

Still, that should happen.

I'm no longer involved with Docker development so I'm gonna remove myself from the mentions. Good luck yall ^_^

**Comment by mmpfeffer** [comment]
The SOC requirement is a good point.  In that case, this feature should be ENABLED with a configuration option to be added to the system-wide docker configuration.  That way SOC requirements can be retained.

**Comment by mmpfeffer** [comment]
For SOC compliance the option must not be enabled.

**Comment by shqear93** [comment]
+1

**Comment by grunzwei** [comment]
> It's only been three years - let's not lose hope now

5.

**Comment by thaJeztah** [comment]
This proposal (in its current form) is very unlikely to be implemented, for various
reasons, among which;

- it keeps the person managing the docker daemon in charge of what connections
  the daemon is allowed to make. note that this option can be "live reloaded",
  so the daemon does not have to be restarted in order to change the configuration.
- every command or code-path that potentially interacts with a registry will have
  to be modified; not just `docker pull`, but also `docker build`, `docker run`,
  `docker plugin`, `docker service` and `docker stack` subcommands, as well as
  orchestrators such as swarmkit, which pull images from worker nodes.
- SOC compliance (as mentioned above)


> At the very least, the docker daemon should be configurable to ALLOW insecure
> registry parameter to be passed in. That moves the security design to the proper
> place - in the hands of the administrator, and outside of docker itself.

I think the administrator in this case would be the person/team managing the
hosts on which the docker daemon runs, not the user that connects to the remote
API. This is the reason for this configuration to be a daemon configuration.

> patching json configs from shell scripts is a major pain point.

That's a fair point, but orthogonal to this discussion. It's not _impossible_
to patch the JSON config, but agreed that it can be more complicated than other
file formats. Note that this configuration can also be set as through flags on
the daemon, which allows you to use a systemd drop-in unit file to reconfigure
the daemon.

> Something like "ALLOW_INSECURE_REGISTY_OPTION" which enables the --insecure-registry flag on the docker command line.

If you want to allow insecure pulls from any registry (which would be the equivalent
of adding an `--unsecure-registry` flag), you can allow "the internet" as insecure
registry; the following should allow any IPv4 address to be used as insecure registry,
(thus fall back to non-TLS connections);


Through the `/etc/docker/daemon.json` configuration file;

```json
{"insecure-registries": ["0.0.0.0/1","128.0.0.0/2","192.0.0.0/3","224.0.0.0/4"]}
```

Or by passing the options as flags on the daemon (which could be set in a systemd
override file);

```bash
dockerd \
    --insecure-registry=0.0.0.0/1 \
    --insecure-registry=128.0.0.0/2 \
    --insecure-registry=192.0.0.0/3 \
    --insecure-registry=224.0.0.0/4
```

**Comment by erfantkerfan** [comment]
+1

**Comment by babinos87** [comment]
Just came across this. We are running a private registry in AWS network, so we need to use internal DNS, when we access through a VPN and want to push images to this registry, it is impossible if we don't edit docker config, which is not ideal for the reasons mentioned by other people. We could go with plain HTTP, but this is even less secure than allowing an insecure registry (which is not so insecure in fact, it's only that the certificate used cannot be validated properly because of hos mismatch!).

Ideally, we would like to specify this extra flag in the docker push command, so no developer has to go through editing the docker configuration. I know it's not big deal, but adds an (unnecessary IMHO) level of complexity.

It seems this is an ongoing topic for many many many years, let's hope that on the issue's 10th anniversary it will have been resolved!

**Comment by andrask** [comment]
Actually, the fundamental issue here is the centralized daemon behavior of Docker. The daemon is not necessarily under the control of the actual user, thus security wise the current setup is probably the best solution. But over the years several alternative toolsets have emerged. For example, with [Skopeo](https://github.com/containers/skopeo) one can easily transfer images between registries. Alternatively, podman can do the same. Therefore, I think it is probably time to conclude this ticket with "rejected" and add notes about the alternatives.

**Comment by adamnovak** [comment]
A workaround for this, since registries on all ports of `localhost` are insecure by default, is to forward a port from `localhost` to the desired registry, pull from `localhost`, and then adjust the tags. So to pull `whatever.tld/imageowner/imagename:tag` from an insecure registry on `whatever.tld` port 80, you would do:

```
REGISTRY_HOST=whatever.tld
IMAGE_ON_REGISTRY=imageowner/imagename:tag
REGISTRY_PORT=80
PROXY_PORT=5050
socat tcp-l:${PROXY_PORT},fork,reuseaddr tcp:${REGISTRY_HOST}:${REGISTRY_PORT} &
docker pull localhost:${PROXY_PORT}/${IMAGE_ON_REGISTRY}
docker tag localhost:${PROXY_PORT}/${IMAGE_ON_REGISTRY} ${REGISTRY_HOST}/${IMAGE_ON_REGISTRY}
docker rmi localhost:${PROXY_PORT}/${IMAGE_ON_REGISTRY}
kill %1
```

**Comment by Martin-Molinero** [comment]
+1
